### PR TITLE
Better paper broker

### PR DESCRIPTION
**Pull request by shershah010**
on `order_limit` and `order_options_limit` set status to open and set status to filled on `fetch_stock_order_status`. Checks if the user can preform the purchase or sale before setting status to filled.

### CODE REVIEW

File: harvest/api/paper.py [hunk 1]
```diff
@@ -113,28 +114,40 @@ def fetch_stock_order_status(self, id: int) -> Dict[str, Any]:
...
-                        'avg_price': price,
-                        'quantity': qty
-                    })
+                # Check to see if user has enough funds to buy the stock
+                if self.buying_power < price * qty:
+                    warning(f"""Not enough buying power.\n Total price ({price} * {qty}) exceeds buying power {self.buy_power}.\n Reduce purchase quantity or increase buying power.""")
+                # Check to see the price does not exceed the limit price
+                elif ret['limit_price'] < price:
+                    limit_price = ret['limit_price']
+                    info(f'Limit price for {sym} is less than current price ({limit_price} < {price}).')
```
Comment (shershah010) [inline harvest/api/paper.py:132]: Check to see that the user can purchase the stocks and the price is lower than the limit_price.

File: harvest/api/paper.py [hunk 2]
```diff
@@ -113,28 +114,40 @@ def fetch_stock_order_status(self, id: int) -> Dict[str, Any]:
...
+                    else:
+                        pos['avg_price'] = (pos['avg_price']*pos['quantity'] + price*qty)/(qty+pos['quantity'])
+                        pos['quantity'] = pos['quantity'] + qty 
+            
+                    self.cash -= price * qty 
+                    self.buying_power -= price * qty 
+                    ret_1 = ret.copy()
+                    self.orders.remove(ret)
+                    ret = ret_1
+                    ret['status'] = 'filled'
```
Comment (shershah010) [inline harvest/api/paper.py:150]: set the status to filled only if the stock was successfully bought.

File: harvest/utils.py [hunk 3]
```diff
@@ -73,6 +73,9 @@ def mark_up(x):
 def mark_down(x):
     return round(x * 0.95, 2)
 
+def has_timezone(date: dt.datetime) -> bool:
+    return date.tzinfo is not None and date.tzinfo.utcoffset(date) is not None
```
Comment (shershah010) [inline harvest/utils.py:77]: helper function to check if datetime object has tz info

File: harvest/wizard.py [hunk 4]
```diff
@@ -64,7 +56,7 @@ def println(self, text:str) -> None:
         print(f"{header}{text}")
         self.text_counter += 1
 
-    def get_bool(self, prompt: str='y/n', true_pat:str=r'y|yes', false_path:str=r'n|no', default:str=None, persistent=True) -> bool:
```
Comment (shershah010) [inline harvest/wizard.py:67]: fixed typo

File: test/test_paper_broker.py [hunk 5]
```diff
@@ -37,7 +37,7 @@ def test_buy_order_limit(self):
         dummy = PaperBroker() 
         dummy.streamer = DummyStreamer()
         dummy.setup(['A'], '1MIN')
-        order = dummy.order_limit('buy', 'A', 5, 25)
+        order = dummy.order_limit('buy', 'A', 5, 50000)
```
Comment (shershah010) [inline test/test_paper_broker.py:40]: set order limit to a high value to make sure that the order gets filled.

File: test/test_wizard.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
+# Builtins
```
Comment (shershah010) [inline test/test_wizard.py:1]: tests for the wizard class.

File: harvest/api/paper.py [hunk 7]
```diff
@@ -19,7 +19,7 @@ class PaperBroker(API):
 
     interval_list = ['1MIN', '5MIN', '15MIN', '30MIN', '1DAY']
 
-    def __init__(self, account_path: str=None):
+    def __init__(self, account_path: str=None, commission_fee: float=0.01):
```
Comment (tfukaza) [inline harvest/api/paper.py:22]: Commission fees can be fixed (like options on [Ameritrade](https://www.tdameritrade.com/pricing.html)), or based on a percentage (like [Gemini](https://www.gemini.com/fees/api-fee-schedule#section-api-fee-schedule)), so ideally there should be a way to switch between that. Additionally, brokers can have different brokerage fees depending on the type of asset.

**Review by tfukaza — CHANGES_REQUESTED** [review]

File: harvest/api/paper.py [hunk 8]
```diff
@@ -19,7 +20,16 @@ class PaperBroker(API):
...
+    def __init__(self, account_path: str=None, commission_fee='1%'):
+        """
+        :commission_fee: When this is a number it is assumed to be a flat price
+            on all buys and sells of assets. When this is a string formatted as
+            'XX%' then it is assumed that commission fees are that percent of the
+            original cost of the buy or sell. When commission fee is a dictionary
+            with the keys 'buy' and 'sell' you can specify different commission 
+            fees when buying and selling assets. The values must be numbers or 
+            strings formatted as 'XX%'.
+        """
```
Comment (shershah010) [inline harvest/api/paper.py:32]: The commission fee now supports flat prices and percentages and can distinguish between selling and buying assets,

File: harvest/api/paper.py [hunk 9]
```diff
@@ -305,4 +331,36 @@ def order_option_limit(self, side: str, symbol: str, quantity: float, limit_pric
...
+        e = 0
+        for asset in self.stocks + self.cryptos + self.options:
+            add = asset['avg_price'] * asset['quantity']
+            if 'multiplier' in asset:
+                add = add * asset['multiplier']
+            e += add
+        e += self.cash
+        return e
 
+    def apply_commission(self, inital_price: float, commission_fee, side: str) -> float:
```
Comment (shershah010) [inline harvest/api/paper.py:350]: applies a commission, the returned value is how much the user actually has to pay or actually receives after the commission is applied

**Comment by tfukaza** [comment]
👍

**Review by tfukaza — APPROVED** [review]

**Pull request merged**